# Analysis Results

## Overview

This notebook analyzes the classification results from Notebook 04.

The previous notebook trained several classical machine learning models on HOG features extracted from Canny sketch images. Logistic Regression was selected as the best model based on validation macro F1-score and was then evaluated on the held-out test set.

This notebook focuses on interpreting the results rather than training new models. It reviews the model comparison, analyzes class-level performance, inspects confusion patterns, and prepares visual examples of correct and incorrect predictions.

The goal is to understand not only how well the model performs, but also where and why it makes mistakes.

## Setup and Imports

In [1]:
from pathlib import Path
import sys

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import cv2

from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    ConfusionMatrixDisplay,
)

sys.path.append("..")

%reload_ext autoreload
%autoreload 2

from src import utils

pd.set_option("display.max_columns", None)
pd.set_option("display.max_colwidth", 120)

plt.rcParams["figure.figsize"] = (10, 6)
plt.rcParams["figure.dpi"] = 100

print("Libraries loaded successfully")

Libraries loaded successfully


## Load Saved Results

The saved outputs from Notebook 04 are loaded here.

These files include the validation model comparison table and the final test predictions for the selected Logistic Regression model. The prediction file contains the relative image path, true label, predicted label, and a correctness flag for each test sample.

In [2]:
tables_dir = utils.OUTPUTS_DIR / "tables"

model_comparison_path = tables_dir / "validation_model_comparison.csv"
test_predictions_path = tables_dir / "test_predictions.csv"

model_comparison_df = pd.read_csv(model_comparison_path)
test_predictions_df = pd.read_csv(test_predictions_path)

print(f"Model comparison shape: {model_comparison_df.shape}")
print(f"Test predictions shape: {test_predictions_df.shape}")

Model comparison shape: (4, 3)
Test predictions shape: (375, 4)


In [3]:
model_comparison_df

,model,accuracy,macro_f1
0,Logistic Regression,0.786667,0.787726
1,Random Forest,0.781333,0.781674
2,SVM,0.765333,0.768859
3,KNN,0.474667,0.473720


In [4]:
test_predictions_df.head()

,relative_path,true_label,predicted_label,correct
0,houzz/dressers/Transitional/1623transitional-dressers.jpg,dresser,dresser,True
1,houzz/sofas/Contemporary/189contemporary-sofas.jpg,sofa,sofa,True
2,houzz/tables/Modern/10919modern-coffee-tables.jpg,table,chair,False
3,houzz/tables/Modern/8865modern-side-tables-and-end-tables.jpg,table,table,True
4,houzz/chairs/Contemporary/2685contemporary-recliner-chairs.jpg,chair,chair,True


In [5]:
print("Model comparison columns:")
print(model_comparison_df.columns.tolist())

print("\nTest prediction columns:")
print(test_predictions_df.columns.tolist())

print(f"\nTotal test predictions: {len(test_predictions_df)}")
print(f"Correct predictions: {test_predictions_df['correct'].sum()}")
print(f"Incorrect predictions: {(~test_predictions_df['correct']).sum()}")

Model comparison columns:
['model', 'accuracy', 'macro_f1']

Test prediction columns:
['relative_path', 'true_label', 'predicted_label', 'correct']

Total test predictions: 375
Correct predictions: 295
Incorrect predictions: 80


### Observation

The saved result files were loaded successfully.

The model comparison table contains the validation accuracy and macro F1-score for the four tested classifiers. The test prediction file contains 375 predictions from the held-out test set, including the true label, predicted label, and correctness flag for each sample.

There are **295 correct predictions** and **80 incorrect predictions**. These incorrect cases will be used later in this notebook to inspect typical model errors.

## Review Model Comparison

## Analyze Final Test Performance

## Class-Level Performance

## Confusion Patterns

## Wrong Prediction Examples

## Correct Prediction Examples

## Key Findings

## Limitations

## Summary and Next Steps